[DebugInfo] Emit debug info for witness tables

augusto2112 opened the pull request:
This change emits debug info for witness tables passed into generic functions when a generic type is constrained to a protocol. This information is required for LLDB's generic expression evaluator to work in such functions.

rdar://104446865

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

augusto2112 commented:
@swift-ci smoke test

slavapestov commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -3711,23 +3720,45 @@ void IRGenDebugInfoImpl::emitTypeMetadata(IRGenFunction &IGF,
+
+void IRGenDebugInfoImpl::emitWitnessTable(IRGenFunction &IGF,
+                                          llvm::Value *Metadata, unsigned Depth,
+                                          unsigned Index,
```
Keying this off of depth/index is insufficient. Eg in the following, we pass a witness table for the `T.Element: Equatable` conformance:

```
func f<T: Sequence>(_: T) where T.Element: Sequence {}
```

augusto2112 committed:
[DebugInfo] Emit debug info for witness tables

This change emits debug info for witness tables passed into generic
functions when a generic type is constrained to a protocol. This
information is required for LLDB's generic expression evaluator
to work in such functions.

rdar://104446865

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@augusto2112
Member
Author
[augusto2112](https://github.com/augusto2112) commented [20 hours ago](https://github.com/swiftlang/swift/pull/79171#issuecomment-2670008951)
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test source compatibility

augusto2112 commented:
I think source compatibility passed, I see this at the end of the logs:

```
[2025-05-06T06:22:39.777Z] Action Summary:
[2025-05-06T06:22:39.777Z]      Passed: 354
[2025-05-06T06:22:39.777Z]      Failed: 0
[2025-05-06T06:22:39.777Z]     XFailed: 39
[2025-05-06T06:22:39.777Z]     UPassed: 0
[2025-05-06T06:22:39.777Z]       Total: 393
[2025-05-06T06:22:39.777Z] ========================================
[2025-05-06T06:22:39.777Z] Repository Summary:
[2025-05-06T06:22:39.777Z]       Total: 144
[2025-05-06T06:22:39.777Z] ========================================
[2025-05-06T06:22:39.777Z] Result: XFAIL
```
And later:
```
[2025-05-06T06:22:55.206Z] + ./swift-source-compat-suite/cleanup main
[2025-05-06T06:22:55.461Z] ** CLEANUP **
[2025-05-06T06:22:55.461Z] Deleting build and source directories...
[2025-05-06T06:28:25.205Z] Cancelling nested steps due to timeout
[2025-05-06T06:28:25.207Z] Sending interrupt signal to process
[2025-05-06T06:28:32.415Z] Traceback (most recent call last):
[2025-05-06T06:28:32.415Z]   File "/Users/ec2-user/jenkins/workspace/swift-PR-source-compat-suite-debug-macos/./swift-source-compat-suite/cleanup", line 95, in <module>
[2025-05-06T06:28:32.415Z]     sys.exit(main())
[2025-05-06T06:28:32.415Z]   File "/Users/ec2-user/jenkins/workspace/swift-PR-source-compat-suite-debug-macos/./swift-source-compat-suite/cleanup", line 80, in main
[2025-05-06T06:28:32.415Z]     common.check_execute([
[2025-05-06T06:28:32.415Z]   File "/Users/ec2-user/jenkins/workspace/swift-PR-source-compat-suite-debug-macos/swift-source-compat-suite/common.py", line 220, in check_execute
[2025-05-06T06:28:32.415Z] /Users/ec2-user/jenkins/workspace/swift-PR-source-compat-suite-debug-macos@tmp/durable-1321e01a/script.sh.copy: line 1: 34317 Terminated: 15          ./swift-source-compat-suite/cleanup main
[2025-05-06T06:28:32.418Z] script returned exit code 143
```

I'll rerun it just in case though

augusto2112 commented:
@swift-ci test source compatibility

augusto2112 commented:
@swift-ci test source compatibility

augusto2112 commented:
Looks like source compatibility aborts because of sccache:
```
20:09:24  + /usr/local/bin/sccache --stop-server
20:09:24  Stopping sccache server...
17:30:55  Sending interrupt signal to process
17:30:55  Aborted by [unknown](https://ci.swift.org/user/unknown)
17:31:02  /Users/ec2-user/jenkins/workspace/swift-PR-source-compat-suite-debug-macos@tmp/durable-fa95a614/script.sh.copy: line 1: 63847 Terminated: 15          /usr/local/bin/sccache --stop-server 2> /dev/null
```

adrian-prantl commented:
Let's merge this and watch the bots closely. Did it pass source-compatibility locally?

augusto2112 commented:
@swift-ci test source compatibility
